src/krate.rs

```diff
@@ -529,37 +534,34 @@ impl Crate {
             },
         };
 
-        // First try to un-delete if they've been soft deleted previously, then
-        // do an insert if that didn't actually affect anything.
-        let amt = conn.execute("UPDATE crate_owners
-                                        SET deleted = FALSE
-                                      WHERE crate_id = $1 AND owner_id = $2
-                                        AND owner_kind = $3",
-                               &[&self.id, &owner.id(), &owner.kind()])?;
-        assert!(amt <= 1);
-        if amt == 0 {
-            conn.execute("INSERT INTO crate_owners
-                               (crate_id, owner_id, created_by, owner_kind)
-                               VALUES ($1, $2, $3, $4)",
-                         &[&self.id, &owner.id(), &req_user.id,
-                             &owner.kind()])?;
-        }
+        let crate_owner = CrateOwner {
+            crate_id: self.id,
+            owner_id: owner.id(),
+            created_by: req_user.id,
+            owner_kind: owner.kind() as i32,
+        };
+        diesel::insert(&crate_owner.on_conflict(
+                crate_owners::table.primary_key(),
+                do_update().set(crate_owners::deleted.eq(false)),
+            )).into(crate_owners::table)
+            .execute(conn)?;
 
         Ok(())
     }
 
     pub fn owner_remove(&self,
-                        conn: &GenericConnection,
+                        conn: &PgConnection,
                         _req_user: &User,
                         login: &str) -> CargoResult<()> {
         let owner = Owner::find_by_login(conn, login).map_err(|_| {
             human(&format_args!("could not find owner with login `{}`", login))
         })?;
-        conn.execute("UPDATE crate_owners
-                              SET deleted = TRUE
-                            WHERE crate_id = $1 AND owner_id = $2
-                              AND owner_kind = $3",
-                     &[&self.id, &owner.id(), &owner.kind()])?;
+        let target = crate_owners::table.find((
+            self.id(),
+            owner.id(),
+            owner.kind() as i32,
+        ));
+        diesel::delete(target).execute(conn)?;
```

Comment:
Oh whoops, you're right this should be soft deleting. I'm not sure why soft delete was implemented either. It seems odd.